pkg/contractreader/observed.go

```diff
@@ -0,0 +1,159 @@
+package contractreader
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+)
+
+var (
+	buckets = []float64{
+		float64(10 * time.Millisecond),
+		float64(20 * time.Millisecond),
+		float64(50 * time.Millisecond),
+		float64(70 * time.Millisecond),
+		float64(100 * time.Millisecond),
+		float64(200 * time.Millisecond),
+		float64(500 * time.Millisecond),
+		float64(700 * time.Millisecond),
+		float64(time.Second),
+		float64(2 * time.Second),
+		float64(5 * time.Second),
+	}
+
+	CrDirectRequestsDurations = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "contract_reader_direct_request_duration",
+			Help:    "The amount of time elapsed during the ChainReader's GetLatestValue execution",
+			Buckets: buckets,
+		},
+		[]string{"chainID", "contract", "method"},
+	)
+	CrBatchRequestsDurations = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "contract_reader_batch_request_duration",
+			Help:    "The amount of time elapsed during the ChainReader's BatchGetLatestValues execution",
+			Buckets: buckets,
+		},
+		[]string{"chainID"},
+	)
+	CrBatchSizes = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "contract_reader_batch_request_size",
+			Help: "The size of the batch request",
+		},
+		[]string{"chainID"},
+	)
+	CrErrors = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "contract_reader_errors",
+			Help: "The number of errors that occurred during the ChainReader's execution",
+		},
+		[]string{"chainID", "function", "contract"},
+	)
+)
+
+type Observed struct {
+	ContractReaderFacade
+	lggr    logger.Logger
+	chainID string
+
+	// Prometheus components for tracking metrics
+	directRequestsDurations *prometheus.HistogramVec
+	batchRequestsDurations  *prometheus.HistogramVec
+	batchSizes              *prometheus.CounterVec
+	errors                  *prometheus.CounterVec
+}
+
+func NewObserverReader(
```

Comment:
We already have `ExtendedContractReader`, could you put this functionality there?

Reply:
`ExtendedContractReader` exposes different methods, it's an adapter on top of ContractReaderFacade. Also, it's not used everywhere so it would require a lot of refactoring now to begin with switching everything to `Extended`.

I want to start with building a simple decorator first that wraps only `ContractReader` interface. At the end of the day, every execution flows through these basic methods. Additionally, if we decide to make it generic it should be as simple as moving this `NewObserverReader` to the chainlink-common without any changes to the code.

Reply:
Thanks, good rational :+1: